staging/src/k8s.io/apiserver/pkg/server/filters/longrunning.go

```diff
@@ -39,3 +40,83 @@ func BasicLongRunningRequestCheck(longRunningVerbs, longRunningSubresources sets
 		return false
 	}
 }
+
+// WithLongRunningRequestTermination starts closing long running requests upon receiving TerminationStartCh signal.
+//
+// It does it by running two additional go routines.
+// One for running the request and the second one for intercepting the termination signal and propagating it to the requests.
+// If the request is not terminated after receiving a signal from inFlightRequestsFinished it will be forcefully killed.
+//
+// This filter exists because sometimes propagating the termination signal is not enough.
+// It turned out that long running requests might block on:
+//  io.Read() for example in https://golang.org/src/net/http/httputil/reverseproxy.go
+//  http2.(*serverConn).writeDataFromHandler which might be actually an issue with the std lib itself
+//
+// Instead of trying to identify current and future issues we provide a filter that ensures terminating long running requests.
+//
+// Also note that upon receiving termination signal the http server sends
+// sends GOAWAY with ErrCodeNo to tell the client we're gracefully shutting down.
+// But the connection isn't closed until all current streams are done.
+func WithLongRunningRequestTermination(handler http.Handler, longRunningFunc apirequest.LongRunningRequestCheck, terminationStartCh <-chan struct{}, inFlightRequestsFinished <-chan struct{}) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
I'm with @sttts that we should be able to get rid of one of the two go-routines.

The exact same model that we're using in the timeout handler should work, no?
```
go func() {
  defer func() {
     err := recover()
     if err != nil {
        ..
     }
     resultCh <- err
  }()
  select {
  case err := <-resultCh:
     ..
  case <-terminationStartCh:
    ..
  }
```

Reply:
This still adds a go-routine per long-running request, so I would like to see some scale test results to check if it's not introducing a visible regression...
@kubernetes/sig-scalability @mborsz @marseel